qiskit/pulse/schedule.py

```diff
@@ -1066,15 +1267,40 @@ def append(
             )
 
         if not inplace:
-            ret_block = copy.deepcopy(self)
-            ret_block._name = name or self.name
-            ret_block.append(block, inplace=True)
-            return ret_block
-        else:
-            self._blocks.append(block)
-            self._parameter_manager.update_parameter_table(block)
+            schedule = copy.deepcopy(self)
+            schedule._name = name or self.name
+            schedule.append(block, inplace=True)
+            return schedule
+
+        if isinstance(block, Reference):
+            self.references[block.ref_keys] = None
+
+        elif isinstance(block, ScheduleBlock):
+            block = copy.deepcopy(block)
+            # Expose subroutines to the current main scope.
+            # Note that this 'block' is not called.
+            # The block is just directly appended to the current scope.
+            if block.is_referenced():
+                if block._parent is not None:
+                    # This is an edge case:
+                    # If this is not a parent, block.references points to the parent's reference
+                    # where subroutine not referred within the 'block' may exist.
+                    # Move only references existing in the 'block'.
+                    # See 'test.python.pulse.test_reference.TestReference.test_appending_child_block'
+                    for ref in _get_references(block._blocks):
+                        self.references[ref.ref_keys] = block.references[ref.ref_keys]
```

Comment:
What if one of these references is already assigned in self?

Reply:
Thanks, done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/670b333f500c276cd174dcfe49f5179f153c7e3c